segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/AggregatedAppendIntegrityChecker.java

```diff
@@ -0,0 +1,180 @@
+/**
+ * Copyright Pravega Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.pravega.segmentstore.server.writer;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import io.pravega.common.util.BufferView;
+import io.pravega.segmentstore.server.DataCorruptionException;
+import io.pravega.segmentstore.server.logs.operations.StreamSegmentAppendOperation;
+import lombok.AccessLevel;
+import lombok.Data;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.util.ArrayDeque;
+import java.util.Iterator;
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Helper class to check internal data integrity of AggregatedAppends at the StorageWriter level before flushing to LTS.
+ * The main purpose of this class is to compute and keep track of hashes of Append contents upon ingestion and then
+ * verify that the integrity of such Appends at the end of the ingestion pipeline (StorageWriter level).
+ */
+@Slf4j
+class AggregatedAppendIntegrityChecker implements AutoCloseable {
+    //region Members
+
+    @GuardedBy("this")
+    @Getter(AccessLevel.PACKAGE)
+    @VisibleForTesting
+    private final Queue<AggregatedAppendIntegrityInfo> appendIntegrityInfo;
+    private final AtomicReference<BufferView> previousPartialAppendData;
```

Comment:
If this is also guarded by `this` then there is no need for it to be atomic. Is that the case?

Reply:
Fixed.